### DISCUSSION TITLE
How do I use Babel's tokenizer?

### DISCUSSION THREAD

**Discussion by samualtnorman** [babel/babel#16989] [Q&A]
I can see that `@babel/parser` has a tokenizer at https://github.com/babel/babel/tree/main/packages/babel-parser/src/tokenizer but I figure out how to access it in Node.js from its NPM package.

Thanks.

**Comment by nicolo-ribaudo** [comment]
You have to use the full parser, passing the `tokens: true` option.

The tokenizer runs interleaved with the parser, getting back info from the parser itself to know how to tokenize the next token, so it's not possible to run it standalone.

**Reply by samualtnorman** [reply]
ah, so if there's invalid syntax, I can't get the tokens?

**Reply by samualtnorman** [reply]
I'm guessing that's yes

**Reply by nicolo-ribaudo** [reply]
Right, yes.

If you only care about _approximate_ tokens that also work for invalid code, you could use https://www.npmjs.com/package/js-tokens (it's what we use in `@babel/code-frame`)

**Reply by samualtnorman** [reply]
I did take a look at that package, but the tokens didn't appear to have source locations
